AP_OSD: Add CRSF link stats display fields

rmaia3d opened the pull request:
As described [here](https://www.expresslrs.org/info/signal-health/), knowing both RSSI and LQ is important to know the health of your radio signal. Knowing the current RSSI in dbm allows comparing to receiver sensitivity limit and knowing when getting there, or close. 

The CRSF protocol (used by both Crossfire and ELRS radio systems) provides a link stats frame with detailed statistics that include this info and a few other items, but there currently is no way to put them on the OSD display of Ardupilot based systems. This PR adds those fields.

I have used as a base the work already done by @shellixyz in implementing these fields for [Arducustom](https://github.com/ArduCustom/ardupilot) and made some changes and improvements.

Main improvement is detecting when BTFL fonts are being used and automatically replace the non-existent icons in the DJI fontset with direct letter representations (in the LQ and Tx power (uplink) fields). It uses more character space, but since the HD grid is larger, it shouldn't be too much of a problem.

Here's how it looks in BTFL fonts mode:
![Ardu_CRSF_LNK_STATS_BTFL_OSD_anno](https://github.com/ArduPilot/ardupilot/assets/9812730/e1bf8dd9-c378-4a70-86be-d85ea5fe1096)

And in analog mode:
![Ardu_CRSF_LNK_STATS_Analog_OSD_anno2](https://github.com/ArduPilot/ardupilot/assets/9812730/a8259d0c-36e4-41db-ae1b-70cb94763c59)

The added fields (RSSI in dBm, Tx uplink power, LQ in % of packets received, uplink SNR in dB and active rx antenna) are highlighted by the red ellipses in the images above.

The fields are individually configurable (no need to use all 5 at once, the user can choose which one they want to use) and also I have added configurable blinking alert levels for the RSSI, SNR and LQ fields.

Hwurzburg commented:
Will review in detail later, but the flash cost for 1M boards is pretty big....we might have to make some this a build option and not default...

rmaia3d commented:
> Will review in detail later, but the flash cost for 1M boards is pretty big....we might have to make some this a build option and not default...

If it helps in that regard, I did all the testing of these changes on two MatekF405 boards (1M flash), both plane and copter (with the copter also running bdshot and IMU batch sampling). :-)

rmaia3d commented:
Build Summary comparison between current master branch and the PR branch:

### Copter:

**- master (1d0fc4e)**
```
Build directory: .../ardupilot/build/MatekF405
Target          Text (B)  Data (B)  BSS (B)  Total Flash Used (B)  Free Flash (B)  External Flash Used (B)
----------------------------------------------------------------------------------------------------------
bin/arducopter    877840      2388   128776                880228          102812  Not Applicable

```
**- ap_osd_crsf_lnk_stats (515717c)**
```
Build directory: .../ardupilot/build/MatekF405
Target          Text (B)  Data (B)  BSS (B)  Total Flash Used (B)  Free Flash (B)  External Flash Used (B)
----------------------------------------------------------------------------------------------------------
bin/arducopter    879076      2388   128780                881464          101572  Not Applicable
```

### Plane:

**- master (1d0fc4e)**
```
Build directory: .../ardupilot/build/MatekF405
Target         Text (B)  Data (B)  BSS (B)  Total Flash Used (B)  Free Flash (B)  External Flash Used (B)
---------------------------------------------------------------------------------------------------------
bin/arduplane    920808      2296   128856                923104           59924  Not Applicable
```

**- ap_osd_crsf_lnk_stats (515717c)**
```
Build directory: .../ardupilot/build/MatekF405
Target         Text (B)  Data (B)  BSS (B)  Total Flash Used (B)  Free Flash (B)  External Flash Used (B)
---------------------------------------------------------------------------------------------------------
bin/arduplane    922428      2296   128860                924724           58300  Not Applicable
```

From these, the total flash impact is from 1.2 to 1.6kB.

Hwurzburg commented:
its in the CI test results....1.6KB is considerable...will need to be discussed in dev call

rmaia3d commented:
You have a point there, I didn't pay attention to this fact...

andyp1per commented:
For analog I don't think this is an improvement - far too much text on the screen. In general I think we are looking for abbreviations that don't take up too much real estate. Also, as Henry says, 1.6Kb is way too much for such a feature.

andyp1per commented:
Also the CRSF specific stuff needs to be hidden behind the CRSF define

rmaia3d commented:
> For analog I don't think this is an improvement - far too much text on the screen. In general I think we are looking for abbreviations that don't take up too much real estate. Also, as Henry says, 1.6Kb is way too much for such a feature.

It is not necessary to put all fields on the screen at once. I only did it to show, in a single image, how they all look. They are individually selectable and configurable.

For analog (and other HD systems that allow custom fonts to be uploaded) @shellixyz has included an extra set of icons in the Arducustom fonts. This greatly reduces the screen real state while retaining full functionality.

![Arducustom_extra_osd_icons](https://github.com/ArduPilot/ardupilot/assets/9812730/4761c881-650f-4cb2-a872-a8045475d01d)

I chose to not include these extra icons in this PR to avoid opening another can of worms. I'm not sure how open the core team would be to expanding that set of files, so I settled for the text based compromise without expanding the icons.

For DJI systems (BTFL fonts enabled) this is moot, because that font/icon set is not customizable outside of DJI itself.

rmaia3d commented:
> Also the CRSF specific stuff needs to be hidden behind the CRSF define

Do you believe this is a feature that could possibly be accepted into the core Ardupilot code? Even as a non-default build option, like @Hwurzburg suggested a few comments back? I understand flash usage is a concern, but I feel there's many people who would like this data on their OSD and don't have the programming expertise to implement it like I did, thus my intention behind this contribution.

I ask this because if there's openness to including, I can rework it under the CRSF define, and see if I can improve the flash footprint as well. But if the core team feels this is really not worth being included, then there's no point in me putting in that time and effort.

Hwurzburg commented:
I think that adding these panels as a build option would be fine...those that really want these CRSF RF extensions could get a version off the server...easy change to this PR..would need crsf dependency in build option,among the other obvious ones...

rmaia3d commented:
That's great! I believe that a build option is a very good compromise. No unnecessary flash usage but, with just a little bit of extra effort, the feature will be there for those who want it. It is fine by me! 

Is this the macro that the CRSF specific code should be reworked under?

`#if AP_RCPROTOCOL_CRSF_ENABLED`

Also, any other macros/defines I should use to separate out the code of this PR?

Hwurzburg commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp:
```diff
@@ -539,7 +544,10 @@ void AP_RCProtocol_CRSF::process_link_stats_frame(const void* data)
         }
     }
 
-    _link_status.rf_mode = MIN(link->rf_mode, 7U);
```
I am sure @andyp1per put this here for a reason....remember this is used with ELRS also

rmaia3d replied:
I removed this cap because ELRS now goes up to RFMD 14, it doesn't stop at 7. But if it being higher than 7 will break other stuff in the code, then I can put the cap back on, no problem. Here's the full table:

<img width="695" alt="Screen Shot 2023-12-29 at 19 35 35" src="https://github.com/ArduPilot/ardupilot/assets/9812730/2b9a6a85-73b8-406d-97e8-8f902b734188">

Source: https://www.expresslrs.org/info/signal-health/#rf-mode-indexes-rfmd

rmaia3d replied:
Ok, in the changes to put the code under the CSRF_EXTENSION wrapper, I have left the rf_mode as is, with the cap at 7.

andyp1per replied:
Its fine to change it to 14. Please but in a #define for the constant so that it's clear what is going on. If you change it you need to change the tables as well to be of the larger size - that's why its constrained to stop reading free memory from the tables

rmaia3d replied:
Ok, thanks for heads up!! Added all the 14 modes, with a define to keep it constrained and also updated the tables. It's all in commit (4c4f167). Let me know if there's anything else I have missed.

Hwurzburg reviewed (changes requested):
you need to ifdef all the OSD panel additions with a #if AP_OSD_CRSF_EXTENSIONS_ENABLED wrapper...and in the header do an #ifndef to disable it, so its only in when specified in the build server  like here https://github.com/ArduPilot/ardupilot/blob/6fb99d6b1ebc5bf3bbd8edef199fd0bc8b34100f/libraries/AP_OSD/AP_OSD_Screen.cpp#L352
- add this opton in build_options.py for the custom build server under Other and make it require CRSF rc option
- add to extract_features.py based on a symbol only present when this feature is compiled in
- probably should ifdef out the added variables also in CRSF RC

rmaia3d commented:
> you need to ifdef all the OSD panel additions with a #if AP_OSD_CRSF_EXTENSIONS_ENABLED wrapper...and in the header do an #ifndef to disable it, so its only in when specified in the build server like here

Ok, thanks a lot for the pointers! I will make these changes and report back as soon as they are done!

rmaia3d commented:
Just added the commits putting the code under the wrappers, so if this isn't chosen amongst the build options, the flash impact is zero. The default is disabled.

Please check and let me know if I have added correctly to the python build scripts (commit f557423).

andyp1per commented:
> > For analog I don't think this is an improvement - far too much text on the screen. In general I think we are looking for abbreviations that don't take up too much real estate. Also, as Henry says, 1.6Kb is way too much for such a feature.
> 
> It is not necessary to put all fields on the screen at once. I only did it to show, in a single image, how they all look. They are individually selectable and configurable.
> 
> For analog (and other HD systems that allow custom fonts to be uploaded) @shellixyz has included an extra set of icons in the Arducustom fonts. This greatly reduces the screen real state while retaining full functionality.
> 
> ![Arducustom_extra_osd_icons](https://private-user-images.githubusercontent.com/9812730/293377215-4761c881-650f-4cb2-a872-a8045475d01d.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQxMjU0NTQsIm5iZiI6MTcwNDEyNTE1NCwicGF0aCI6Ii85ODEyNzMwLzI5MzM3NzIxNS00NzYxYzg4MS02NTBmLTRjYjItYTg3Mi1hODA0NTQ3NWQwMWQuUE5HP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDFUMTYwNTU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDNiOGE4NjQ4MWNkMTU5YWNkODljZWM0MDE1Mzc5MTIzMTgzNjc0NzgzZWE1OTRmNWFkYjJlODA3YjdmYzM2YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.fCgBP2LrQhoyrGvq-I1yS6WRmfJhm9DwW8AeaDx8FoU)
> 
> I chose to not include these extra icons in this PR to avoid opening another can of worms. I'm not sure how open the core team would be to expanding that set of files, so I settled for the text based compromise without expanding the icons.
> 
> For DJI systems (BTFL fonts enabled) this is moot, because that font/icon set is not customizable outside of DJI itself.

This is a much better solution - I think you should include the commits from ArduCustom here to support this.

andyp1per commented:
> > Also the CRSF specific stuff needs to be hidden behind the CRSF define
> 
> Do you believe this is a feature that could possibly be accepted into the core Ardupilot code? Even as a non-default build option, like @Hwurzburg suggested a few comments back? I understand flash usage is a concern, but I feel there's many people who would like this data on their OSD and don't have the programming expertise to implement it like I did, thus my intention behind this contribution.
> 
> I ask this because if there's openness to including, I can rework it under the CRSF define, and see if I can improve the flash footprint as well. But if the core team feels this is really not worth being included, then there's no point in me putting in that time and effort.

I think what you need to do is build some support for this in the forums. I think if 10 people say this is great then that's a very different calculation to just you. I let some stuff of this nature go into CRSF in the past and was never very happy with the outcome - most people found the output spammy and indeed the utility is quite subjective. Obviously having it completely compiled out by default makes it less intrusive, but I would rather we got this right from the get go than having to live with something sub-optimal. I do know that people like having both RSSI db and LQ in CRSF, so I am receptive to the overall idea.

rmaia3d commented:
> This is a much better solution - I think you should include the commits from ArduCustom here to support this.

Yes, I agree it is best solution so far. I have copied over the fonts from Arducustom, including all of the added symbols there. There's an extra 16 symbols, not all of them being used by this PR, but it does leave room for even more improvements and customizations in the future.

The new font files and the code additions to support them are in commits e9e75cc and cc58599.

However, these additions and the ELRS RFMD tables expansion (in commit 4c4f167) are not under the CRSF_EXTENSIONS or any other define, they would be global to all builds. Combined, including the new font files, this results in a 200 byte increase in flash usage.

Here's how the updated icons look in analog OSD (second image shows with RFMD prefixing the LQ value, added via commit d2ca7ee):

![Ardu_CRSF_LNK_STATS_Analog_OSD_CustSymb_PureLQ](https://github.com/ArduPilot/ardupilot/assets/9812730/7a9630bd-cfbd-4603-a4f1-def76e1a0439) 
![Ardu_CRSF_LNK_STATS_Analog_OSD_CustSymb_RFMDpreLQ](https://github.com/ArduPilot/ardupilot/assets/9812730/44cea7f2-943f-4671-81b4-c7f614218412)

rmaia3d commented:
> I think what you need to do is build some support for this in the forums. I think if 10 people say this is great then that's a very different calculation to just you. I let some stuff of this nature go into CRSF in the past and was never very happy with the outcome - most people found the output spammy and indeed the utility is quite subjective. Obviously having it completely compiled out by default makes it less intrusive, but I would rather we got this right from the get go than having to live with something sub-optimal. I do know that people like having both RSSI db and LQ in CRSF, so I am receptive to the overall idea.

Ok, that sounds fair! Initially, the idea was indeed to just implement the LQ and RSSI dbm panels, however I realized the work involved in implementing all of them wasn't that much, so it was a "why not" kind of decision.

Meanwhile, I have added further commits with more improvements (described in the previous comment) and also some code reorganizations for improved readability and organization. 

I have separated out the methods for drawing the new panels under BTFL fonts mode, where there aren't convenient symbols for labelling the panels. So there's the "normal" method which uses the custom symbols (available in analog and HD systems like WS and HDZero) and there's the _btfl method which uses direct letter representations without any custom symbols (which only DJI could implement). The choice between one method or the other is automatically done based on the BTFL fonts OSD setting being enabled or not.

Since the HD screen has a lot more pixel real-state, I believe the space used by the extra characters (instead of the single character icons) is not that bad and is a good compromise. The extra methods increased a little the flash footprint, but in my opinion the improved organization and tidiness is worth it.

andyp1per commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -85,7 +85,11 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
     // @Param: _OPTIONS
     // @DisplayName: OSD Options
     // @Description: This sets options that change the display
+#if OSD_CRSF_PANELS_ENABLED
```
I think we now prefer an AP prefix so AP_OSD_CRSF_PANELS_ENABLED

andyp1per replied:
I think you should just unilaterally add the docs, they will at least then show up in mission planner. We do this in other places as well.

rmaia3d replied:
Ok, I have done these changes in the latest commits.

andyp1per commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -112,12 +116,35 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
+    // @Description: Set level at which RSSI item will flash (-dBm)
+    // @Range: -128 0
+    // @User: Standard
+    AP_GROUPINFO("_W_RSSI", 12, AP_OSD, warn_rssi, -100),
```
RSSI is a standard metric not limited to CRSF, your PR should allow this feature to be used elsewhere

andyp1per replied:
I think you have 11 characters free here (so OSDx_ + addition), so you should try and make the name more meaningful. _WARN_RSSI would fit for instance

andyp1per replied:
Rather than change the whole parameter I think just change the default in a #define somewhere which is what you are looking for

andyp1per commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -112,12 +116,35 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
+    // @Description: Set level at which CRSF_LQ item will flash (%)
+    // @Range: 0 100
+    // @User: Standard
+    AP_GROUPINFO("_W_LQ", 33, AP_OSD, warn_lq, 50),
```
ditto name

rmaia3d replied:
I haven't changed this to keep consistency. All other warning level parameters follow this "_W_" pattern, so I chose to keep it that way.

andyp1per commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -112,12 +116,35 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
+    // @Description: Set level at which CRSF_SNR item will flash (in db)
+    // @Range: -20 10
+    // @User: Standard
+    AP_GROUPINFO("_W_SNR", 34, AP_OSD, warn_snr, 0),
```
ditto name

rmaia3d replied:
Same as above.

andyp1per commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -239,7 +266,7 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
     // @User: Standard
     // @RebootRequired: True
-    AP_GROUPINFO("_TYPE2", 32, AP_OSD, osd_type2, 0),
+    AP_GROUPINFO("_TYPE2", 32, AP_OSD, osd_type2, 0),    
```
remove whitespace

andyp1per commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -1077,6 +1082,89 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info2[] = {
+    // @DisplayName: CRSFPWR_Y
+    // @Description: Vertical position on screen
+    // @Range: 0 21
+    AP_SUBGROUPINFO(crsf_tx_power, "CRSFPWR", 55, AP_OSD_Screen, AP_OSD_Setting),
```
ditto comment on naming

rmaia3d replied:
I have added an underline/underscore to most of these fields, except the CRSFRSSI panel, as that would make it go above the 16 char limit on OSDx_CRSF_RSSI_EN (this is 17 char long) and create havoc.

andyp1per commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -1917,6 +2023,192 @@ void AP_OSD_Screen::draw_esc_amps(uint8_t x, uint8_t y)
+            backend->write(x, y, blink, "LQ--:--");
+        }        
+        else {            
+            backend->write(x, y, blink, "LQ%2d:%2d", rf_mode, value);
```
A chunk of the flash cost is all these very similar functions. I think you should be able to boil this down to a lookup table and a much smaller set of functions

rmaia3d replied:
Indeed! I have rearranged these functions (in commit [d419b69](https://github.com/ArduPilot/ardupilot/pull/25844/commits/d419b699c89347baeca8474349063f18be8caa4f)) to the best I could. A lookup table isn't practical due to the variadic arguments in the calls to `backend->write()` and the requirement of the format string to be a `const char*`. I couldn't think of any solution that made sense other than the traditional "if else" structuring. But they are now all contained inside the relevant functions, without duplicates. Saved about 200 bytes of flash.

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp:
```diff
@@ -523,8 +528,9 @@ void AP_RCProtocol_CRSF::process_link_stats_frame(const void* data)
     } else {
         rssi_dbm = link->uplink_rssi_ant2;
-    }
+    }        
```
remove whitespace

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp:
```diff
@@ -537,9 +543,19 @@ void AP_RCProtocol_CRSF::process_link_stats_frame(const void* data)
-    _link_status.rf_mode = MIN(link->rf_mode, 7U);
+    }    
+
+    _link_status.rf_mode = MIN(link->rf_mode, ELRS_MAX_RFMD_MODES); // Cap to avoid memory spills in the conversion tables
```
It would probably be better to define this as RFMode::RF_MODE_MAX_MODES - RFMode::CRSF_MAX_MODES

rmaia3d replied:
Done! And I also removed the 4Hz setting from ELRS, as that seems to not be used at all in ELRS. Instead, I renamed that field to CRSF_RF_MAX_MODES. It's in [5b6ddcb](https://github.com/ArduPilot/ardupilot/pull/25844/commits/5b6ddcb3fab536f5001b69e2b87a2ad68350220d)

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -252,6 +252,9 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
     };
 
+    // Limit the number of ELRS RF modes based on: https://www.expresslrs.org/info/signal-health/#rf-mode-indexes-rfmd
+    # define ELRS_MAX_RFMD_MODES 14
```
see comment above

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -252,6 +252,9 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
+
     enum RFMode {
         CRSF_RF_MODE_4HZ = 0,
         CRSF_RF_MODE_50HZ,
```
add in CRSF_MAX_MODES here and then you can use it in the ELRS calculation

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -261,19 +264,43 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
+        int16_t rssi = -1;
+        int16_t link_quality = -1;
+        int8_t rf_mode = -1;
+        int16_t tx_power = -1;
```
put the extra stuff inside the existing definition with a #define rather than side-by-side

andyp1per reviewed (changes requested):
Looking better! Some comments added

Hwurzburg commented on Tools/scripts/build_options.py:
```diff
@@ -309,6 +309,7 @@ def __init__(self,
     Feature('Other', 'SDCARD_MISSION', 'AP_SDCARD_STORAGE_ENABLED', 'Enable storing mission on microSD cards', 0, None),
     Feature('Other', 'COMPASS_CAL', 'COMPASS_CAL_ENABLED', 'Enable "tumble" compass calibration', 0, None),
     Feature('Other', 'DRONECAN_SERIAL', 'AP_DRONECAN_SERIAL_ENABLED', 'Enable DroneCAN virtual serial ports', 0, None),
+    Feature('Other', 'OSD_CRSF_LINK_STATS', 'AP_OSD_CRSF_EXTENSIONS_ENABLED', 'Enable OSD panels for CRSF link stats data', 0, "RC_CRSF"),
```
```suggestion
    Feature('Other', 'OSD_CRSF_LINK_STATS', 'AP_OSD_CRSF_EXTENSIONS_ENABLED', 'Enable OSD panels for CRSF expanded link stats data', 0, "RC_CRSF"),
```

rmaia3d replied:
Thanks! Good catch!

Hwurzburg commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -114,10 +114,26 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
     // @DisplayName: RSSI warn level (in %)
-    // @Description: Set level at which RSSI item will flash
-    // @Range: 0 99
+    // @Description: Set level at which RSSI item will flash (in % or dBm if applicable)
```
```suggestion
    // @Description: Set level at which RSSI item will flash (in positive % or negative dBm values as applicable). 30% or -100dBm are defaults.
```

rmaia3d replied:
Another great idea! Will update like this.

Hwurzburg reviewed (changes requested):
Also commits need to be squashed and Tools/gittools/got-subsystems-split used

rmaia3d commented:
> Also commits need to be squashed and Tools/gittools/got-subsystems-split used

Do you mean squash everything into a single commit and then run the git-subsystems-split tool? Is there any reference for usage of this script?

Hwurzburg commented:
> > Also commits need to be squashed and Tools/gittools/got-subsystems-split used
> 
> Do you mean squash everything into a single commit and then run the git-subsystems-split tool? Is there any reference for usage of this script?

yes....its not is the wiki, but I will add it to the contributing code section....I dont speed a lot of time on the dev wiki since all my time is consumed on making the user side more friendly....and it could use some attention...

rmaia3d commented:
Ok, so I created a local test branch to experiment on and this is the result I got after squashing all the 22 commits in this PR to a single one and then running the git-subsystems-split script:

![Ardu_CRSF_LNK_STATS_git-subsystems-split_output_01](https://github.com/ArduPilot/ardupilot/assets/9812730/9196829b-2bbd-41ae-855c-8dcd9de70c33)

After that, git log --oneline outputs this:

![Ardu_CRSF_LNK_STATS_git-subsystems-split_output_02](https://github.com/ArduPilot/ardupilot/assets/9812730/00e3c9ce-8740-48f3-8361-86cc68ce38d9)

Is this the expected outcome?

Hwurzburg commented:
yep...but you need to fix the merge conflict in extract features, and do the squash and split again

rmaia3d commented:
> yep...but you need to fix the merge conflict in extract features, and do the squash and split again

Ok, thanks! I have rebased my branch to the current state of master (as of now), fixed the conflicts, squashed and split. :-)

andyp1per commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -552,6 +588,9 @@ class AP_OSD
         OPTION_BF_ARROWS = 1U<<5,
         OPTION_AVIATION_AH = 1U<<6,
+#if AP_OSD_CRSF_PANELS_ENABLED
+        OPTION_RF_MODE_ALONG_WITH_LQ = 1U << 20,
```
Don't you mean 1U<<7 here?

rmaia3d replied:
Good catch!! Yeah, 7 makes a lot more sense than 20. Fixed it.

andyp1per commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -1077,6 +1082,89 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info2[] = {
+    // @DisplayName: CRSFRSSI_Y
+    // @Description: Vertical position on screen
+    // @Range: 0 21
+    AP_SUBGROUPINFO(crsf_rssi_dbm, "CRSFRSSI", 54, AP_OSD_Screen, AP_OSD_Setting),
```
I tend to think we should not lose the _ here. Either all without or with. I guess you could always use XF intead of CRSF

rmaia3d replied:
Yes, keeping the _ consistent across all panels is much better. I have changed the prefixes to XF so we are able to keep the underscore on all names. But I have changed on the parameters only, everything else (defines, variable names, etc, remain "crsf").

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -252,29 +252,50 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
         CRSF_RF_MODE_50HZ,
         CRSF_RF_MODE_150HZ,
         CRSF_RF_MODE_250HZ,
-        ELRS_RF_MODE_4HZ,
```
You need both here, just assign them the same numeric value

rmaia3d replied:
Ok, thanks for the heads-up. Added both with the same numeric value.

andyp1per reviewed (changes requested):
Looking good! A few small things to fix

rmaia3d commented:
> Looking good! A few small things to fix

Great!! I have just committed the fixes you mentioned. 

Please let me know if you feel like there are further changes needed so I can add them before doing the squash/split process with the commits.

Hwurzburg commented on Tools/scripts/build_options.py:
```diff
@@ -311,6 +311,7 @@ def __init__(self,
     Feature('Other', 'SDCARD_MISSION', 'AP_SDCARD_STORAGE_ENABLED', 'Enable storing mission on microSD cards', 0, None),
     Feature('Other', 'COMPASS_CAL', 'COMPASS_CAL_ENABLED', 'Enable "tumble" compass calibration', 0, None),
     Feature('Other', 'DRONECAN_SERIAL', 'AP_DRONECAN_SERIAL_ENABLED', 'Enable DroneCAN virtual serial ports', 0, None),
+    Feature('Other', 'OSD_CRSF_LINK_STATS', 'AP_OSD_CRSF_EXTENSIONS_ENABLED', 'Enable OSD panels for CRSF expanded link stats data', 0, "RC_CRSF"),
```
should be in OSD section and needs OSD as a dependency in addition to RC_CRSF

rmaia3d replied:
Done!!

rmaia3d commented:
I have just pushed another commit changing the parameter id numbers to be consistent with the sequence inside the GroupInfo array.

rmaia3d commented:
Rebased to current state of master, fixed the merge conflicts in the python scripts, squashed and split the commits.

andyp1per commented on libraries/AP_OSD/fonts/HDFonts/WS/font_update.ini:
```diff
@@ -0,0 +1,11 @@
+1=user_ardu
+
+[user_ardu]
+imgname_720=user_ardu_24.png
```
These are interesting - how are they used?

rmaia3d replied:
These files are specific to the Walksanail digital video system, and they should go into the goggles (or VRX) SD card font folder. I decided to include them for completeness, as the files that come "standard" with the system for Ardupilot don't contain the extra set of icons implemented in this PR. But the system accepts custom fonts (unlike DJI), so it's just a matter of replacing the "stock" files with these customized ones. It doesn't affect the actual Ardupilot code at all.

andyp1per commented on Tools/scripts/extract_features.py:
```diff
@@ -242,6 +242,7 @@ def __init__(self, filename, nm="arm-none-eabi-nm", strings="strings"):
             ('FORCE_APJ_DEFAULT_PARAMETERS', 'AP_Param::param_defaults_data'),
             ('HAL_BUTTON_ENABLED', 'AP_Button::update'),
             ('HAL_LOGGING_ENABLED', 'AP_Logger::Init'),
+            ('AP_OSD_CRSF_EXTENSIONS_ENABLED', r'AP_OSD_Screen::draw_crsf_tx_power'),
```
This name is different to the PANELS name - why is that?

andyp1per replied:
Seems like you have two defines - they both should go in here

rmaia3d replied:
Yes, there are two defines. The AP_OSD_CRSF_EXTENSIONS_ENABLED is supposed to be the "global" one, that can be combined with other "globals" to enable the panels themselves via a "local" define (the AP_OSD_CRSF_PANELS_ENABLED). On AP_OSD.h, line 55 [here](https://github.com/ArduPilot/ardupilot/pull/25844/files#diff-2d27ad6b9c73dbb0c2e46daf08ef2b34117a0665902d9115d1185d53e61e52aeR55) is where the PANELS define is defined and it's value set based on both globals that it should be dependent on. 

I decided to do it this way (with the two defines) to make it easier to manage multiple dependencies, if needed.

andyp1per commented on Tools/scripts/build_options.py:
```diff
@@ -110,6 +110,7 @@ def __init__(self,
     Feature('OSD', 'PLUSCODE', 'HAL_PLUSCODE_ENABLE', 'Enable PlusCode', 0, 'OSD'),
     Feature('OSD', 'OSD_PARAM', 'OSD_PARAM_ENABLED', 'Enable OSD param', 0, 'OSD'),
     Feature('OSD', 'OSD_SIDEBARS', 'HAL_OSD_SIDEBAR_ENABLE', 'Enable Scrolling Sidebars', 0, 'OSD'),
+    Feature('OSD', 'OSD_CRSF_LINK_STATS', 'AP_OSD_CRSF_EXTENSIONS_ENABLED', 'Enable OSD panels for CRSF expanded link stats data', 0, "OSD,RC_CRSF"),
```
need the PANELS define as well I think

rmaia3d replied:
Since the PANELS define is conditionally defined [here](https://github.com/ArduPilot/ardupilot/pull/25844/files#diff-2d27ad6b9c73dbb0c2e46daf08ef2b34117a0665902d9115d1185d53e61e52aeR55), as explained in the previous reply, I don't think it's necessary to add it to build_options, as it's supposed to be "local" define to the AP_OSD code.

Hwurzburg replied:
if you provide an ability to exclude something, you need to be able to selectively build it if its removed...otherwise remove the exclusion if it will never be used

rmaia3d replied:
> if you provide an ability to exclude something, you need to be able to selectively build it if its removed...otherwise remove the exclusion if it will never be used

Check AP_OSD.h, line 55 [here](https://github.com/ArduPilot/ardupilot/pull/25844/files#diff-2d27ad6b9c73dbb0c2e46daf08ef2b34117a0665902d9115d1185d53e61e52aeR55), there you will see what I mean.

tridge commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -1917,6 +2023,154 @@ void AP_OSD_Screen::draw_esc_amps(uint8_t x, uint8_t y)
+    {
+        if(tx_power < 1000) {
+            if(btfl) {
+                backend->write(x, y, false, "%3dMW", tx_power);
```
should me mW not MW, as its milliWatts not MegaWatts

rmaia3d replied:
I have made a quick change to use the 0x0C character instead of "M", and this is how it looks like:

![IMG_8101](https://github.com/ArduPilot/ardupilot/assets/9812730/1dd2ca53-a2ce-424f-8e98-d7fec01ec2b6)

Acceptable?

tridge commented:
@andyp1per will rebase and test

rmaia3d commented:
> @andyp1per will rebase and test

Ok, great!! Thanks!

andyp1per commented:
@rmaia3d can you do the rebase?

rmaia3d commented:
> @rmaia3d can you do the rebase?

Yes, done, no problem!

I have also included the change to use a smaller "m" (from the available DJI fontset) in the BTFL mW representation, as @tridge commented.

Instead of "MW", this is how it looks like:

![Ardu_CRSF_LNK_STATS_0x0C_mW_change](https://github.com/ArduPilot/ardupilot/assets/9812730/434dc0d0-fcd8-4344-bb82-42dd81a72195)

andyp1per commented:
@rmaia3d you need to fix build options to pass the python format check

rmaia3d commented:
> @rmaia3d you need to fix build options to pass the python format check

I see it's complaining about the line being longer than 127 characters, but there are other lines even longer in that file... This line, for example, is 172 characters long (the one I added is 150 chars long). Only difference is the `# NOQA: E501` comment at the end. 

![image](https://github.com/ArduPilot/ardupilot/assets/9812730/1fbca252-07e0-4724-86fa-e1d07ad10389)

There are other longer lines with that same comment as well. What's that supposed to mean? Should I add it too?

I ask because the obvious fix would be simply breaking into two separate lines after one of the arguments, but I see that this isn't done anywhere in the build_options.py file.

andyp1per commented:
i think add the comment and the warning will go away

rmaia3d commented:
Yep, that did it! :-)

andyp1per commented:
Also RSSI seems to fluctuate a lot - not sure whether that's expected

rmaia3d commented:
> The new font is great! But I don't appear to be getting units on many of the XF settings. This is on Walksnail - does it have all the characters?

Did you load the custom Walksnail fonts in libraries/AP_OSD/fonts/HDFonts/WS into your goggles SD card? The fonts that come by default in the goggles are based on the "original" analog fonts for Ardu, which don't include these units. You need either those fonts or the "ArduCustom" fonts from SneakyFPV: 

https://sites.google.com/view/sneaky-fpv/home

Hopefully, after this has been merged into the main Ardupilot codebase, the default Walksnail fonts will be updated so there will be no need anymore to load a custom font (only if really desired).

rmaia3d commented:
> Also RSSI seems to fluctuate a lot - not sure whether that's expected

Yes, that's normal. Raw dbm varies a lot, specially if the craft is changing orientation constantly and flying close by.

andyp1per commented:
> Did you load the custom Walksnail fonts in libraries/AP_OSD/fonts/HDFonts/WS into your goggles SD card?

Yes I did

andyp1per commented:
> Yes, that's normal. Raw dbm varies a lot, specially if the craft is changing orientation constantly and flying close by.

Sitting in my office not moving, TX not moving...

andyp1per commented:
Can you post a picture of what each _XF_ setting should look like in the new font scheme so that I can verify? Not sure this is quite working right. Do you have WS?

rmaia3d commented:
> Can you post a picture of what each _XF_ setting should look like in the new font scheme so that I can verify? Not sure this is quite working right. Do you have WS?

There you go:

![image](https://github.com/ArduPilot/ardupilot/assets/9812730/6589dfee-1af7-41fa-bf45-d8a0ca8e309c)

And here inserted into a "complete" OSD (CRSF itens are pointed to by the blue arrows):

![image](https://github.com/ArduPilot/ardupilot/assets/9812730/cad018a1-df06-49bd-9d00-352f067fa499)

rmaia3d commented:
For reference, the step by step process of adding custom fonts to the Walksnail system:

First, copy the font png files to the "userfont" folder on the SD card you use on your WS goggles or VRX:

![image](https://github.com/ArduPilot/ardupilot/assets/9812730/4e61f926-bc91-4b8b-96cd-a4041b766877)

Next, open the "font_update.ini" file in a text editor. On the top section under the `[config]` header, add another entry with the name of the font, as pointed by the blue arrow (this is how it's going to be displayed in the goggles/vrx menu), Make sure the leading number is not duplicated in any other entry and also is correctly ordered (pointed to by the green arrow). Also, update the `count=` item with the total amount of fonts available (pointed to by the purple arrow):

![image](https://github.com/ArduPilot/ardupilot/assets/9812730/0a1d31fb-982d-43d5-b54d-f3727cb9a39e)

With the "font_update.ini" file still open, scroll to the end of it and add the font configuration section. You can copy and paste from an existing entry. Just make sure to change the header title (between brackets, pointed by the blue arrow) to the same name as you added on the previous step. Then change the `imgname` entries (purple arrows) to the same filenames you copied into the SD card "userfont" folder in the first step (note that there's a file for 720p resolution and another one for 1080p).

![image](https://github.com/ArduPilot/ardupilot/assets/9812730/9f289a2c-5007-4e00-8375-94a3e75764b3)

Now close the text editor, reinsert the SD card into the Goggles/VRX, power it up and open the menu. Navigate to the "Settings -> Display" page. Select `Custom OSD` to "Custom" (magenta arrow). Then, on `Custom Font` (yellow arrow), select the font by the name you've given on the steps above.

![image](https://github.com/ArduPilot/ardupilot/assets/9812730/6c711b1c-8686-4eee-8f82-ebe394742454)

Voilá, enjoy the new fonts! :)

andyp1per commented:
Ok, I had the fonts installed correctly, but you need to pick the "Custom" font rather than the "ARDU" font. I don't get the naming like you do - so maybe that's a difference in the latest firmware.

Ok, so the remaining issue I now see is that some of the symbols that ship with the default ARDU font set are much nicer than the ones in that directory. Here are the fonts that were already on my googles - you can see that many of the symbols are richer and higher quality than the ones replacing them:

![WS_ARDU_Europa_24](https://github.com/ArduPilot/ardupilot/assets/2893260/2868f8f1-9af8-483e-a1e3-8d1e28a0271a)
![WS_ARDU_Europa_36](https://github.com/ArduPilot/ardupilot/assets/2893260/43ab4768-b2cf-4910-9786-b2036af0a704)

rmaia3d commented:
@andyp1per Yes, indeed, the fonts you posted look much nicer!

They are SneakyFpv's version, and he also has the extended fonts:

![image](https://github.com/ArduPilot/ardupilot/assets/9812730/173ac115-77be-468e-992f-99174c675834)
 
(I cropped the file just to highlight the extra symbols available)

Full files are available from here (the file is SNEAKY_FPV_WS_APC_Europa.zip):
https://drive.google.com/drive/folders/1wTU11HyU0JC7kEd-jJ8B9exN_LXrH4Eo

We could include these fonts with this PR instead, I am not against as I agree it's much richer and detailed.

I will just check with SneakyFPV if he's ok with it, after all, it's his work.

rmaia3d commented:
> I will just check with SneakyFPV if he's ok with it, after all, it's his work.

After speaking with SneakyFPV, it is ok to include his fonts here. Once this PR gets merged and becomes "official", there is even the possibility that these new and improved fonts get included in the goggles (at least Walksnail) by default in future firmwares, which is also great. :-)

@andypiper, have you made more testing? Did you find out anything else that needs changing?

andyp1per commented:
Can you add the font files and I will retest. Your PR needs a rebase

andyp1per commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -114,10 +114,26 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
-    AP_GROUPINFO("_W_RSSI", 12, AP_OSD, warn_rssi, 30),
+    AP_GROUPINFO("_W_RSSI", 12, AP_OSD, warn_rssi, AP_OSD_WARN_RSSI_DEFAULT),
+
+#if AP_OSD_CRSF_PANELS_ENABLED
```
Please move these to the bottom - it makes it very hard for future editors to add new parameters if they don't know that some of these indexes are used

rmaia3d replied:
Ok, done!!! 

Indeed, good catch to make life easier for future changes. :-)

andyp1per commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -226,6 +237,15 @@ class AP_OSD_Screen : public AP_OSD_AbstractScreen
 
+#if AP_OSD_CRSF_PANELS_ENABLED
+    // CRSF link stats data panels
+    AP_OSD_Setting crsf_tx_power{false, 0, 0};
```
Sorry, one more - please can you provide appropriate default positions here so that if someone enables them they show up in a sensible place in the panel.

rmaia3d replied:
No problem! Done! :-)

andyp1per commented:
Nice work!

rmackay9 commented:
Could you rebase this on master?  We can then review again on the EU dev call.

peterbarker commented on libraries/AP_RCProtocol/AP_RCProtocol_config.h:
```diff
@@ -79,3 +79,6 @@
 #define AP_RCPROTOCOL_MAVLINK_RADIO_ENABLED AP_RCPROTOCOL_BACKEND_DEFAULT_ENABLED && BOARD_FLASH_SIZE > 1024 && HAL_GCS_ENABLED
 #endif
 
+#ifndef AP_OSD_CRSF_EXTENSIONS_ENABLED
```
Shouldn't this be in `AP_OSD/AP_OSD_config.h` ?

peterbarker replied:
If you *really* want it in `AP_RCProtoco_config.h` then you can rename it to `AP_RCPROTOCOL_CRSG_OSD_EXTENSIONS_ENABLED`.

But I do believe it's better to be in `AP_OSD_config.h`; you can include `AP_RCProtocol_config.h` to get `AP_RCPROTOCOL_CRSF_ENABLED`.

Additionally, while at the moment it might depend on the AP_RCProtocol library, this is actually probably not the correct structure.  You can see this in the CRSF "stand-alone" mode where we only process telemetry data, not consume RC, from the CRSF serial stream.  Presumably these OSD extensions would be one use case for that.  Our current structure has the RCProtocol library process the stream, a better structure *may* be to have a separate library that parses the serial stream and pokes data into AP_RCProtocol, or into the OSD library or whatever as appropriate.  This is a long-term project, but I think shows that your assertion that this is primarily an "RC" feature is incorrect.

rmaia3d replied:
Indeed, under that notion of separating the telemetry from the actual RC commands stream my assertion is incorrect. However, I made that assertion based on how things are currently implemented. But yes, you totally have a point that this really depends on the telemetry stream, being independent from the RC commands.

I will look into making this change as you suggested, then I would include `AP_OSD_config.h` into the `AP_RCProtocol_CRSF` files instead of the other way around. And remove the define from `AP_RCProtocol_config.h`.

rmaia3d commented:
> Could you rebase this on master? We can then review again on the EU dev call.

Yes, no problem! Just did it.

rmackay9 commented:
There's a user request [here](https://discuss.ardupilot.org/t/copter-4-5-0-beta4-available-for-beta-testing/115149/7) to backport this to 4.5.x.

rmaia3d commented:
> There's a user request [here](https://discuss.ardupilot.org/t/copter-4-5-0-beta4-available-for-beta-testing/115149/7) to backport this to 4.5.x.

I am 100% ok with that! Just let me know if there's anything I need to do for it (open another PR maybe, since it would be merging into a different branch? Don't know).

peterbarker commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -48,8 +49,18 @@ class AP_MSP;
+#define AP_OSD_NUM_SYMBOLS 107
 #define OSD_MAX_INSTANCES 2
+
+#if AP_OSD_CRSF_EXTENSIONS_ENABLED && AP_RCPROTOCOL_CRSF_ENABLED
```
```suggestion
#if AP_OSD_CRSF_EXTENSIONS_ENABLED
```

Far better to get a compilation failure if someone forgets the define than to silently fail.

The custom build server will take care of adding the extra define based on the dependencies you've specified in `build_options.py`

rmaia3d replied:
But what if someone doesn't use the custom build server, for example building locally from a git repo clone? Wouldn't it be better to leave both conditions there to ensure the correct dependencies are set?

rmaia3d replied:
Since I have renamed the main define, I took the opportunity and removed the second condition, as you suggested.

peterbarker commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -1923,6 +2029,154 @@ void AP_OSD_Screen::draw_esc_amps(uint8_t x, uint8_t y)
+            } else {
+                backend->write(x, y, blink, "LQ--");
+            }
+        }
+        else {
```
```suggestion
        } else {
```

similarly everywhere else

rmaia3d replied:
Ok, thanks for the head up. This was probably VSCode auto-formatting getting me off-guard. Will look into correcting this according to the style guide.

peterbarker commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -1923,6 +2029,154 @@ void AP_OSD_Screen::draw_esc_amps(uint8_t x, uint8_t y)
+    bool btfl = is_btfl_fonts();
+    bool prefix_rf = check_option(AP_OSD::OPTION_RF_MODE_ALONG_WITH_LQ);
+    const int16_t rf_mode = AP::crsf()->get_link_status().rf_mode;    
+    if (lqv < 0)
+    {
```
```suggestion
    if (lqv < 0) {
```

peterbarker commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp:
```diff
@@ -599,7 +605,28 @@ void AP_RCProtocol_CRSF::process_link_stats_frame(const void* data)
+    // Populate the extra data items
+    if(link->uplink_status > 0) {
+        _link_status.rssi_dbm = rssi_dbm;
+        _link_status.tx_power = link->uplink_tx_power < sizeof(AP_RCProtocol_CRSF::tx_powers)
+                                    ? AP_RCProtocol_CRSF::tx_powers[link->uplink_tx_power]
+                                    : -1;
```
```suggestion
        _link_status.tx_power = -1;
        if (link->uplink_tx_power < sizeof(AP_RCProtocol_CRSF::tx_powers))
            _link_status.tx_power = AP_RCProtocol_CRSF::tx_powers[link->uplink_tx_power];
        }
```

(or use an `else`; in either case the ternary operator here make things less readable)

peterbarker replied:
... you could consider doing the "set all fields to -1" as a first step and then filling valid values in with non-minus-one values, actually.

rmaia3d replied:
Good suggestions! Thanks!

peterbarker reviewed (changes requested):
Out style guide is here: https://ardupilot.org/dev/docs/style-guide.html

peterbarker commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -1923,6 +2029,154 @@ void AP_OSD_Screen::draw_esc_amps(uint8_t x, uint8_t y)
+bool AP_OSD_Screen::is_btfl_fonts()
+{
+    const AP_MSP *p_msp = AP::msp();
+    return (p_msp && p_msp->is_option_enabled(AP_MSP::Option::DISPLAYPORT_BTFL_SYMBOLS));
```
```suggestion
    return (p_msp !=nullptr && p_msp->is_option_enabled(AP_MSP::Option::DISPLAYPORT_BTFL_SYMBOLS));
```

rmaia3d replied:
Good one too!

peterbarker commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -1077,6 +1082,89 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info2[] = {
+
+    // @Param: XF_ANT_EN
+    // @DisplayName: XF_ANT_EN
+    // @Description: Displays the current active antenna for CRSF
```
```suggestion
    // @Description: Displays the current active antenna
```

similarly elsewhere.

peterbarker reviewed (dismissed):
There's a few bits in here which are sub-optimal.  Properly, the `rssi_dbm` should go via the `AP_RSSI` library so other places can take advantage of it.

Nothing about the new data coming in (e.g. antenna) is CRSF-specific, so we probably shouldn't be conveying it in a CRSF-specific object.  Why add `XF` to all the new parameters?  Those would be equally valid regardless of where the data comes from.

rmaia3d commented:
> There's a few bits in here which are sub-optimal. Properly, the `rssi_dbm` should go via the `AP_RSSI` library so other places can take advantage of it.
> 
> Nothing about the new data coming in (e.g. antenna) is CRSF-specific, so we probably shouldn't be conveying it in a CRSF-specific object. Why add `XF` to all the new parameters? Those would be equally valid regardless of where the data comes from.

In other firmwares (BetaFlight and Inav), it is implied that this is CRSF specific, so this is "where I'm coming from". Not saying Ardupilot should stick to this, not all. But I am not familiar nor have experience with other systems that report such data. Again, not saying there aren't such systems, just that I have zero knowledge or experience with them.

How about this PR gets accepted in it's current general form, being CRSF specific, and then further work can be built upon that? This would also open to other people contributing to and improving the initial work made here, so it does not depend entirely on myself.

The "XF" is added to all the new parameters mostly to distinguish them from the other, already existing, elements, namely RSSI and LQ. But this can easily be revised to already be non-CRSF specific. For example:
- XF_RSSI could be RSSIDBM (without the underscore to not go over the 16 char limit in the full parameter names)
- XF_SNR could be RC_SNR (RC alluding to rc system)
- XF_ANT could be RC_ANT (same allusion)
- XF_PWR could be TX_PWR (it is the RC system transmit power) or RC_PWR 
- XF_LQ could be RC_LQ

tridge commented:
> * XF_SNR could be RC_SNR (RC alluding to rc system)
> 
> * XF_ANT could be RC_ANT (same allusion)
> 
> * XF_PWR could be TX_PWR (it is the RC system transmit power) or RC_PWR
> 
> * XF_LQ could be RC_LQ

sounds good!
we're not so concerned about the variable names, it is just the param names, so when another R/C system comes along that can provide the same sorts of info we don't need to rename the user facing parameters
thanks!

rmaia3d commented:
> sounds good! we're not so concerned about the variable names, it is just the param names, so when another R/C system comes along that can provide the same sorts of info we don't need to rename the user facing parameters thanks!

Ok!!

I just pushed several changes. I renamed all user facing parameters and also removed the `CRSF` (or `crsf`) specific references in function names, variables, etc, at the AP_OSD side. Indeed, like this it makes a lot more sense going forward.

I have also renamed the main define to `AP_OSD_LINK_STATS_EXTENSIONS_ENABLED` (instead of AP_OSD_CRSF_EXTENSIONS_ENABLED) and updated the python files accordingly. And moved this define to `AP_OSD_Config.h`. 

I also did a few touchups and fixes to the parameter definitions and descriptions, as well as making the style fixes to comply with the style guide. And a few other changes as suggested by @peterbarker.

andyp1per commented:
LGTM

lida2003 commented:
@rmaia3d If I wanna try on tag 4.4.4, what steps should I do?

a) cherry-pick this commit
b) define `AP_OSD_LINK_STATS_EXTENSIONS_ENABLED ` in hwdef,  is there any other macro needed?
c) build a custom binary for update

am i right?

rmaia3d commented:
> @rmaia3d If I wanna try on tag 4.4.4, what steps should I do?
> 
> a) cherry-pick this commit b) define `AP_OSD_LINK_STATS_EXTENSIONS_ENABLED ` in hwdef, is there any other macro needed? c) build a custom binary for update
> 
> am i right?

Yes, cherry-picking would be the easier way. Cherry-pick all 4 commits and it should do it. You may encounter some merge conflicts in the process, but easily resolved.

For item b) adding the define to the hwdef file works, it is one way to do it. Or you can also change in `AP_OSD_Config.h` the define value from `0` to `1`. This will make the extra OSD panels available for all boards when building the custom binary, without the need to change each hwdef file.

For c), yes, you will need to build locally a custom binary and upload it to the board. If you don't already have a local build environment set up, this [docs page](https://ardupilot.org/dev/docs/building-the-code.html) has detailed instructions.

peterbarker commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -214,6 +214,22 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
     // @User: Standard
     AP_GROUPINFO("_W_ACRVOLT", 31, AP_OSD, warn_avgcellrestvolt, 3.6f),
 
+#if AP_OSD_EXTENDED_LNK_STATS
```
```suggestion
#if AP_OSD_EXTENDED_LNK_STATS_ENABLED
```

We'll make this change in a future PR....

rmaia3d replied:
Do you mean you think this is a typo or you'd rather have the define renamed?

rmaia3d commented:
> There's a user request [here](https://discuss.ardupilot.org/t/copter-4-5-0-beta4-available-for-beta-testing/115149/7) to backport this to 4.5.x.

Should I create another PR backporting this to the 4.5.x branch or this will be handled by the dev team without the need of another PR from me?

rmackay9 commented:
@rmaia3d,

I've added it to the 4.5 project so it will be considered for backporting.  If it can be backported to the 4.5 without merge conflicts then no need for a new PR.  I'll usually ask for help if it can't be backported.  Txs!

rmackay9 commented:
I did not add this to 4.5.2-beta1 due to a merge conflict

rmaia3d commented:
> I did not add this to 4.5.2-beta1 due to a merge conflict

Ok, I tried adding these changes to the current Plane-4.5 branch and indeed some merge conflicts showed up in two files (AP_RCProtocol_config.h and extract_features.py). All of them related to features that are present in the 4.6 codebase, but not in the 4.5. I simply removed the "offending" lines and was able to merge and build successfully. 

Here's the branch with the merged changed, on my account. Would you like me to open a PR into Plane-4.5 from this branch?

https://github.com/rmaia3d/ardupilot/tree/pr/4.5/osd_lnk_stats

andyp1per commented:
Yes open a PR I think

rmaia3d commented:
> Yes open a PR I think

Ok!! Done, #26912

lida2003 commented:
> I did not add this to 4.5.2-beta1 due to a merge conflict

Hi @rmaia3d @andyp1per @rmackay9  Currently 4.5.2 doesn't has this feature right? Cause I didn't find those macros:`AP_OSD_EXTENDED_LNK_STATS` `AP_OSD_EXTENDED_LNK_STATS_ENABLED`

rmaia3d commented:
> > I did not add this to 4.5.2-beta1 due to a merge conflict
> 
> Hi @rmaia3d @andyp1per @rmackay9 Currently Copter-4.5.2 doesn't have this feature right? Cause I didn't find those macros:`AP_OSD_EXTENDED_LNK_STATS` `AP_OSD_EXTENDED_LNK_STATS_ENABLED`

Yes, you are correct. It hasn't been included in 4.5.2 but should be in 4.5.3.

andyp1per commented:
@rmaia3d there appears to be a problem with the WS font which is part of this PR - the armed and disarmed symbols appear to be the wrong way round. When disarmed I get a little flying prop symbol and when armed, nothing.

Hwurzburg commented:
Thats the normal way of showing armed/disarmed....same with AP OSDs....additional symbol following mode when disarmed goes away when armed

rmaia3d commented:
> @rmaia3d there appears to be a problem with the WS font which is part of this PR - the armed and disarmed symbols appear to be the wrong way round. When disarmed I get a little flying prop symbol and when armed, nothing.

As @Hwurzburg has said, it looks like this is indeed the normal and expected behavior.

Looking through the code, the `SYM_ARMED` icon is `0x00`, which both on the analog and Walksnail fonts maps to a blank space (also in BTFL fonts mode). When disarmed (`SYM_DISARMED`), the symbol changes, becoming a "framed" D in the analog fonts and the little prop with a red line going over it in Walksnail (in BTFL mode, for the DJI O3 system, it is mapped to a character that resembles an asterisk).

This DISarmed icon in Walksnail from the fonts included in this PR is this one:

<img width="41" alt="Screen Shot 2024-05-21 at 01 43 03" src="https://github.com/ArduPilot/ardupilot/assets/9812730/fa7e6c36-6e97-4ffe-aaae-918541653a7d">

rmaia3d commented:
> Anyway to get 4:100?

@lida2003 Yes! You need to go into the full parameter list, under the `OSD_OPTIONS` parameter and enable this option (which is bit 7):

![image](https://github.com/ArduPilot/ardupilot/assets/9812730/fa73cb2a-d316-474d-b4cb-17d78a204071)

Then the `RC_LQ` panel will display as X:Y where X is the RF Mode number and Y is the actual LQ %age level. For example, 4:100.
